### PR TITLE
Lazy Lists

### PR DESCRIPTION
**Pull request by CodaFi**

### CODE REVIEW

File: Basis/Array.swift [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+public func cons<T>(x : T) -> [T] -> [T] {
+	return { xs in x <| xs }
+}
+
+public func <|<T>(lhs : T, var rhs : [T]) -> [T] {
+	rhs.insert(lhs, atIndex: 0)
+	return rhs
+}
+
```
Comment (pthariensflame) [inline Basis/Array.swift]: Why not have `snoc` (with synonym `|>`) as well?

File: Basis/Search.swift [hunk 2]
```diff
@@ -35,8 +45,23 @@ public func lookup<A : Equatable, B>(e : A) -> [A:B] -> Optional<B> {
...
+			case .Cons(let (x, y), let xys):
+				if e == x {
+					return .Some(y)
+				}
+				return lookup(e)(xys)
+		}
+	}
+}
+
+/// Looks up a key in a dictionary.
```
Comment (pthariensflame) [inline Basis/Search.swift]: I think this comment should be switched with one earlier in the file.  :P

File: Basis/Array.swift (hunk 1 above)
Comment (CodaFi) [inline Basis/Array.swift]: To be honest I don't like overloading the flag operators in the first place.  I'm open to new ideas.

Comment (pthariensflame) [inline Basis/Array.swift]: "Flag operators"?

Comment (CodaFi) [inline Basis/Array.swift]: <| and |>

Comment (pthariensflame) [inline Basis/Array.swift]: How about `«` and `»`?  :þ

Honestly, though, `|>` and `<|` are correct no matter how you interpret them in this case. `cons` and `snoc` are the proper specializations of the one-handed versions of `Alternative`'s `<|>`.

Comment (CodaFi) [inline Basis/Array.swift]: That's what I thought too (except the applicative thing.  I had no idea that's where these came from!), but it worries me that I've also taken a page from elm and used them for function application.  Maybe it's those operators that have to change.

Comment (pthariensflame) [inline Basis/Array.swift]: Ah, I didn't think about that.  Yeah, I thin those should change in Basis (Swiftz is a potentially different matter.).  Might I suggest `~>` and `<~` for explicit function application, then?  (These would the become `>~>` and `<~<` for pre- and post-composition, and they fit well with a more general category/arrow framework.)

Comment (CodaFi) [inline Basis/Array.swift]: If we're talking categories, then <<< and >>> are certainly open.

Comment (CodaFi) [inline Basis/Array.swift]: How about just throwing another > onto them?

```swift
public func |>>
public func <<|
```

Comment (pthariensflame) [inline Basis/Array.swift]: That actually fits, in a kind of meta-pattern way, with my usage of those exact operators in Aquifer.  Go for it!

Comment (CodaFi) [inline Basis/Array.swift]: So it has come to pass.

File: Basis/Operators.swift [hunk 3]
```diff
@@ -44,9 +44,21 @@ infix operator <| {
...
+	associativity left
+}
+
+/// Cons | Constructs a list by appending a given element to the front of a list.
+infix operator <| {
+	precedence 100
+	associativity right
+}
+
+/// Snoc | Constructs a list by appending a given element to the end of a list.
```
Comment (CodaFi) [inline Basis/Operators.swift]: If this the proper associativity?  I've only ever seen snoclists in Idris and they seemed to work well infixl.

Comment (pthariensflame) [inline Basis/Operators.swift]: This should be fine.  The occasional paretheses aren't a dealbreaker for most people, just so long as the precedence is in the right ballpark.

And yes, if you think about it, snoc can *only* work `infixl`.  If you were to try to use it `infixr`, you'd get type errors all over the place!

Comment (CodaFi) [inline Basis/Operators.swift]: Thanks.

The precedence comment is a reminder about the function application stuff.  They should have lowest possible precedence.

**Comment by CodaFi** [comment]
:boat:

**Pull request merged**